mypyc/test-data/irbuild-basic.test

```diff
@@ -3546,3 +3546,420 @@ L0:
+L4:
+L5:
+    return s
+
+[case testForMapUnpack]
+def f(x: int, y: int) -> tuple[int, int]:
+    return (x, y)
+def g(a: list[int], b: list[int]) -> int:
+    s = 0
+    for x, y in map(f, a, b):
+        s += x + y
+    return s
+[out]
+def f(x, y):
+    x, y :: int
+    r0 :: tuple[int, int]
+L0:
+    r0 = (x, y)
+    return r0
+def g(a, b):
+    a, b :: list
+    s :: int
+    r0 :: dict
+    r1 :: str
+    r2 :: object
+    r3, r4, r5 :: native_int
+    r6 :: bit
+    r7 :: native_int
+    r8 :: bit
+    r9 :: object
+    r10, _mypyc_map_arg_0 :: int
+    r11 :: object
+    r12, _mypyc_map_arg_1 :: int
+    r13, r14 :: object
+    r15, r16 :: int
+    r17 :: tuple[int, int]
+    r18, x, r19, y, r20, r21 :: int
+    r22, r23 :: native_int
+L0:
+    s = 0
+    r0 = __main__.globals :: static
+    r1 = 'f'
+    r2 = CPyDict_GetItem(r0, r1)
+    r3 = 0
+    r4 = 0
+L1:
+    r5 = var_object_size a
+    r6 = r3 < r5 :: signed
+    if r6 goto L2 else goto L5 :: bool
+L2:
+    r7 = var_object_size b
+    r8 = r4 < r7 :: signed
+    if r8 goto L3 else goto L5 :: bool
+L3:
+    r9 = list_get_item_unsafe a, r3
+    r10 = unbox(int, r9)
+    _mypyc_map_arg_0 = r10
+    r11 = list_get_item_unsafe b, r4
+    r12 = unbox(int, r11)
+    _mypyc_map_arg_1 = r12
+    r13 = box(int, _mypyc_map_arg_0)
+    r14 = box(int, _mypyc_map_arg_1)
+    r15 = unbox(int, r13)
+    r16 = unbox(int, r14)
+    r17 = f(r15, r16)
+    r18 = r17[0]
+    x = r18
+    r19 = r17[1]
+    y = r19
+    r20 = CPyTagged_Add(x, y)
+    r21 = CPyTagged_Add(s, r20)
+    s = r21
+L4:
+    r22 = r3 + 1
+    r3 = r22
+    r23 = r4 + 1
+    r4 = r23
+    goto L1
+L5:
+L6:
+    return s
+
+[case testForMapComprehension]
+def f(x: int, y: int) -> int:
+    return x * y
+def g(a: list[int], b: list[int]) -> list[int]:
+    return [x for x in map(f, a, b)]
+[out]
+def f(x, y):
+    x, y, r0 :: int
+L0:
+    r0 = CPyTagged_Multiply(x, y)
+    return r0
+def g(a, b):
+    a, b, r0 :: list
+    r1 :: dict
+    r2 :: str
+    r3 :: object
+    r4, r5, r6 :: native_int
+    r7 :: bit
+    r8 :: native_int
+    r9 :: bit
+    r10 :: object
+    r11, _mypyc_map_arg_0 :: int
+    r12 :: object
+    r13, _mypyc_map_arg_1 :: int
+    r14, r15 :: object
+    r16, r17, r18, x :: int
+    r19 :: object
+    r20 :: i32
+    r21 :: bit
+    r22, r23 :: native_int
+L0:
+    r0 = PyList_New(0)
+    r1 = __main__.globals :: static
+    r2 = 'f'
+    r3 = CPyDict_GetItem(r1, r2)
+    r4 = 0
+    r5 = 0
+L1:
+    r6 = var_object_size a
+    r7 = r4 < r6 :: signed
+    if r7 goto L2 else goto L5 :: bool
+L2:
+    r8 = var_object_size b
+    r9 = r5 < r8 :: signed
+    if r9 goto L3 else goto L5 :: bool
+L3:
+    r10 = list_get_item_unsafe a, r4
+    r11 = unbox(int, r10)
+    _mypyc_map_arg_0 = r11
+    r12 = list_get_item_unsafe b, r5
+    r13 = unbox(int, r12)
+    _mypyc_map_arg_1 = r13
+    r14 = box(int, _mypyc_map_arg_0)
+    r15 = box(int, _mypyc_map_arg_1)
+    r16 = unbox(int, r14)
+    r17 = unbox(int, r15)
```

Comment:
The boxing followed by unboxing seems pointless. It would be great if we didn't need this (it might be a pre-existing issue though).

Reply:
Does mypyc currently perform any optimization over the IR before generating the actual C code? I looked briefly at the transform module, which seems like it might play some role here, but I couldn't figure out the entrypoint, or if that section even works as I think it does

Reply:
After looking, I'm not entirely sure what causes this.

The fake NameExpr that I created is using the correct rtype in the IR, and from reading the code in for_helpers I can't figure out why (or where) r14 thru r17 are being emitted. 

What I see is that we create a CallExpr using the NameExpr directly as the input, which implies to me there wouldn't be any intermediate values involved in the actual call.